pkg/util/apiserver/apiserver.go

```diff
@@ -0,0 +1,220 @@
+		as.logger.WithField("groupversion", groupVersion).WithField("pattern", pattern).Info("Adding Resource Handler")
+	}
+
+	// discovery resource
+	list.APIResources = append(as.resourceList[groupVersion].APIResources, resource)
+
+	// add specific crd resource handler
+	key := fmt.Sprintf("%s/%s", groupVersion, resource.Name)
+	as.delegates[key] = handler
+
+	as.logger.WithField("groupversion", groupVersion).WithField("apiresource", resource).Info("Adding APIResource")
+}
+
+// resourceHandler handles namespaced resource calls, and sends them to the appropriate CRDHandler delegate
+func (as *APIServer) resourceHandler(gv string) https.ErrorHandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) error {
+		namespace, resource, err := splitNameSpaceResource(r.URL.Path)
+		if err != nil {
+			https.FourZeroFour(as.logger.WithError(err), w, r)
+			return nil
+		}
+
+		delegate, ok := as.delegates[fmt.Sprintf("%s/%s", gv, resource)]
+		if !ok {
+			https.FourZeroFour(as.logger, w, r)
+			return nil
+		}
+
+		err = delegate(w, r, namespace)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+}
+
+// addSerializedHandler sets up a handler than will send the serialised content
+// to the specified path.
+func (as *APIServer) addSerializedHandler(pattern string, m k8sruntime.Object) {
+	as.mux.HandleFunc(pattern, https.ErrorHTTPHandler(as.logger, func(w http.ResponseWriter, r *http.Request) error {
+		if r.Method == http.MethodGet {
+			info, err := AcceptedSerializer(r, Codecs)
+			if err != nil {
+				return err
+			}
+
+			w.Header().Set(ContentTypeHeader, info.MediaType)
+			err = Codecs.EncoderForVersion(info.Serializer, unversionedVersion).Encode(m, w)
+			if err != nil {
+				return errors.New("error marshalling")
+			}
+		} else {
+			https.FourZeroFour(as.logger, w, r)
+		}
+
+		return nil
+	}))
+}
+
+// AcceptedSerializer takes the request, and returns a serialiser (if it exists)
+// for the given codec factory and
+// for the Accepted media types.  If not found, returns error
+func AcceptedSerializer(r *http.Request, codecs serializer.CodecFactory) (k8sruntime.SerializerInfo, error) {
+	// this is so we know what we can accept
+	mediaTypes := codecs.SupportedMediaTypes()
+	alternatives := make([]string, len(mediaTypes))
+	for i, media := range mediaTypes {
+		alternatives[i] = media.MediaType
+	}
+	header := r.Header.Get(AcceptHeader)
+	accept := goautoneg.Negotiate(header, alternatives)
+	if len(accept) == 0 {
+		accept = k8sruntime.ContentTypeJSON
+	}
+	info, ok := k8sruntime.SerializerInfoForMediaType(mediaTypes, accept)
+	if !ok {
+		return info, errors.Errorf("Could not find serializer for Accept: %s", header)
+	}
+
+	return info, nil
+}
+
+// splitNameSpaceResource returns the namespace and the type of resource
+func splitNameSpaceResource(path string) (namespace, resource string, err error) {
+	list := strings.Split(strings.Trim(path, "/"), "/")
```

Comment:
interesting way to use default value with returned variable.

Reply:
You know, I can't remember if I did this deliberately :thinking:

Reply:
I don't see any issue, it just gave me idea for my own projects :)